llama_index/evaluation/batch_runner.py

```diff
@@ -1,88 +1,328 @@
+            v = eval_kwargs_lists[k]
+            if not isinstance(v, list):
+                raise ValueError(
+                    f"Each value in eval_kwargs must be a list. Got {k}: {v}"
+                )
+            eval_kwargs_lists[k] = self._validate_and_clean_inputs(v)[0]
 
         # run evaluations
         eval_jobs = []
-        for query, response in zip(queries, responses):
+        for idx, query in enumerate(cast(List[str], queries)):
+            response_str = cast(List, response_strs)[idx]
+            contexts = cast(List, contexts_list)[idx]
+            eval_kwargs = self._get_eval_kwargs(eval_kwargs_lists, idx)
             for name, evaluator in self.evaluators.items():
                 eval_jobs.append(
                     eval_worker(
-                        evaluator, name, query, response, query_kwargs.get(query, {})
+                        self.semaphore,
+                        evaluator,
+                        name,
+                        query=query,
+                        response_str=response_str,
+                        contexts=contexts,
+                        eval_kwargs=eval_kwargs,
                     )
                 )
-        results = await asyncio.gather(*eval_jobs)
+        results = await self.asyncio_mod.gather(*eval_jobs)
 
         # Format results
-        results_dict: Dict[str, List[EvaluationResult]] = {
-            name: [] for name in self.evaluators.keys()
-        }
-        for name, result in results:
-            results_dict[name].append(result)
+        return self._format_results(results)
 
-        return results_dict
+    async def aevaluate_responses(
+        self,
+        queries: Optional[List[str]] = None,
+        responses: Optional[List[Response]] = None,
+        **eval_kwargs_lists: Dict[str, Any],
+    ) -> Dict[str, List[EvaluationResult]]:
+        """Evaluate query, response pairs.
+
+        This evaluates queries and response objects.
+
+        Args:
+            queries (Optional[List[str]]): List of query strings. Defaults to None.
+            responses (Optional[List[Response]]): List of response objects.
+                Defaults to None.
+            **eval_kwargs_lists (Dict[str, Any]): Dict of lists of kwargs to
+                pass to evaluator. Defaults to None.
+
+        """
+        queries, responses = self._validate_and_clean_inputs(queries, responses)
+        for k in eval_kwargs_lists.keys():
+            v = eval_kwargs_lists[k]
+            if not isinstance(v, list):
+                raise ValueError(
+                    f"Each value in eval_kwargs must be a list. Got {k}: {v}"
+                )
+            eval_kwargs_lists[k] = self._validate_and_clean_inputs(v)[0]
+
+        # run evaluations
+        eval_jobs = []
+        for idx, query in enumerate(cast(List[str], queries)):
+            response = cast(List, responses)[idx]
+            eval_kwargs = self._get_eval_kwargs(eval_kwargs_lists, idx)
```

Comment:
Oh yeah it would look something like this:
```
eval_results = await batch_runner.aevaluate_response_strs(
    eval_qs, response_strs=pred_response_strs, reference=ref_response_strs
)
```
where ref_response_strs is just a list of strings. Basically it would pass **{"reference": r} into the evaluator call for a single query, where r is just an entry in ref_response_strs.

I agree this is maybe a little confusing...honestly asking as an alternative would you prefer eval_kwargs to be a list of dicts? (that was the other thing I was exploring). Or the original implementation? 

The reason I found the original implementation a bit awk was because specifying the query_str as a key felt a heavy vs. specifying query_id, but then specifying a query_id itself felt like additional steps